### PR TITLE
Implement hit testing API

### PR DESCRIPTION
**Pull request by Manishearth**
Depends on https://github.com/servo/webxr/pull/149 , https://github.com/servo/servo/pull/26170

This implements non-transient hit tests.

The tests that do not pass are due to https://github.com/web-platform-tests/wpt/issues/22898 , https://github.com/web-platform-tests/wpt/issues/22900, https://github.com/web-platform-tests/wpt/issues/22901 , and https://github.com/immersive-web/hit-test/issues/86

### CODE REVIEW

**Comment by Manishearth** [comment]
I haven't gone through all the tests but the reason we fail the refSpace one, for example, is both because of  immersive-web/hit-test#86 and because of another bug in the test where it's assuming hit test results get duplicated

**Comment by Manishearth** [comment]
Rebased over test updates that I landed upstream early.

**Review by asajeffrey — APPROVED** [review]
Mostly LGTM, we're generating a lot of objects per frame, but hey ho.

One thing I don't understand is how to do hit tests relative to a moving refererence, e.g. from the users's hands or eyes? Does this require generating a new ray each frame?

File: components/script/dom/xrframe.rs [hunk 1]
```diff
@@ -92,17 +98,27 @@ impl XRFrameMethods for XRFrame {
...
         Ok(Some(XRPose::new(&self.global(), pose)))
     }
+
+    /// https://immersive-web.github.io/hit-test/#dom-xrframe-gethittestresults
+    fn GetHitTestResults(&self, source: &XRHitTestSource) -> Vec<DomRoot<XRHitTestResult>> {
+        self.data
+            .hit_test_results
+            .iter()
+            .filter(|r| r.id == source.id())
+            .map(|r| XRHitTestResult::new(&self.global(), *r, self))
```
[the remark this answers is not in the document]
Comment (Manishearth) [reply components/script/dom/xrframe.rs:121]: I don't see any other way to do it? Note that it's possible for a single hit test to have multiple results.

File: components/script/dom/xrhittestresult.rs [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+        )
+    }
+}
+
+impl XRHitTestResultMethods for XRHitTestResult {
+    // https://immersive-web.github.io/hit-test/#dom-xrhittestresult-getpose
+    fn GetPose(&self, base: &XRSpace) -> Option<DomRoot<XRPose>> {
+        let base = self.frame.get_pose(base)?;
+        let pose = base.inverse().pre_transform(&self.result.space);
+        Some(XRPose::new(&self.global(), pose.cast_unit()))
```
[the remark this answers is not in the document]
Comment (Manishearth) [reply components/script/dom/xrhittestresult.rs:49]: We already do this with getPose in views

**Comment by CYBAI** [comment]
Needs a `./mach fmt` 👀

```
Diff in /repo/components/script/dom/xrsession.rs at line 747:
             .granted_features()
             .iter()
             .find(|f| &**f == "hit-test")
-            .is_none() {
(B+            .is_none()
(B+        {
(B             p.reject_error(Error::NotSupported);
             return p;
         }
clang-format not installed. Skipping CPP formatting.
Run `./mach fmt` to fix the formatting
```

**Comment by Manishearth** [comment]
Strangely I do not get that error locally

**Comment by Manishearth** [comment]
That patch is also already applied here and on github. I suspect tc is doing something weird.

(tc won't pass anyway until servo/webxr#149 lands)

**Comment by CYBAI** [comment]
```
Checking files for tidiness...
./Cargo.lock:1: duplicate versions for package `autocfg`
	The following packages depend on version 0.1.7 from 'crates.io':
	The following packages depend on version 1.0.0 from 'crates.io':
./components/script/dom/webidls/XRHitTestResult.webidl:10: no newline at EOF
```

**Comment by Manishearth** [comment]
Ah, our webxr tests need to be updated for supportedModes

**Pull request merged**